AsyncDisplayKit/ASVideoPlayerNode.mm

```diff
@@ -0,0 +1,729 @@
+//
+//  ASVideoPlayerNode.m
+//  AsyncDisplayKit
+//
+//  Created by Erekle on 5/6/16.
+//  Copyright © 2016 Facebook. All rights reserved.
+//
+
+#import "ASVideoPlayerNode.h"
+#import "ASDefaultPlaybackButton.h"
+
+static void *ASVideoPlayerNodeContext = &ASVideoPlayerNodeContext;
+
+@interface ASVideoPlayerNode() <ASVideoNodeDelegate>
+{
+  ASDN::RecursiveMutex _videoPlayerLock;
+
+  __weak id<ASVideoPlayerNodeDelegate> _delegate;
+
+  struct {
+    unsigned int delegateNeededControls:1;
+    unsigned int delegateSpinnerTintColor:1;
+    unsigned int delegatePlaybackButtonTint:1;
+    unsigned int delegateScrubberMaximumTrackTintColor:1;
+    unsigned int delegateScrubberMinimumTrackTintColor:1;
+    unsigned int delegateScrubberThumbTintColor:1;
+    unsigned int delegateScrubberThumbImage:1;
+    unsigned int delegateTimeLabelAttributes:1;
+    unsigned int delegateTimeLabelAttributedString:1;
+    unsigned int delegateLayoutSpecForControls:1;
+    unsigned int delegateVideoNodeDidPlayToTime:1;
+    unsigned int delegateVideoNodeWillChangeState:1;
+    unsigned int delegateVideoNodeShouldChangeState:1;
+    unsigned int delegateVideoNodePlaybackDidFinish:1;
+    unsigned int delegateDidTapVideoPlayerNode:1;
+  } _delegateFlags;
+  
+  NSURL *_url;
+  AVAsset *_asset;
+  
+  ASVideoNode *_videoNode;
+
+  NSArray *_neededControls;
+
+  NSMutableDictionary *_cachedControls;
+
+  ASDefaultPlaybackButton *_playbackButtonNode;
+  ASTextNode  *_elapsedTextNode;
+  ASTextNode  *_durationTextNode;
+  ASDisplayNode *_scrubberNode;
+  ASStackLayoutSpec *_controlFlexGrowSpacerSpec;
+  ASDisplayNode *_spinnerNode;
+
+  BOOL _isSeeking;
+  CMTime _duration;
+
+  BOOL _disableControls;
+
+  BOOL _shouldAutoplay;
+  BOOL _shouldAutorepeat;
+  BOOL _muted;
+  int32_t _periodicTimeObserverTimescale;
+  NSString *_gravity;
+
+  BOOL _shouldAggressivelyRecoverFromStall;
+
+  UIColor *_defaultControlsColor;
+}
+
+@end
+
+@implementation ASVideoPlayerNode
+- (instancetype)init
+{
+  if (!(self = [super init])) {
+    return nil;
+  }
+
+  [self _init];
+
+  return self;
+}
+
+- (instancetype)initWithUrl:(NSURL*)url
+{
+  if (!(self = [super init])) {
+    return nil;
+  }
+  
+  _url = url;
+  _asset = [AVAsset assetWithURL:_url];
+  
+  [self _init];
+  
+  return self;
+}
+
+- (instancetype)initWithAsset:(AVAsset *)asset
+{
+  if (!(self = [super init])) {
+    return nil;
+  }
+
+  _asset = asset;
+  _disableControls = NO;
+
+  [self _init];
+
+  return self;
+}
+
+- (void)_init
+{
+
+  _defaultControlsColor = [UIColor whiteColor];
+  _cachedControls = [[NSMutableDictionary alloc] init];
+
+  _videoNode = [[ASVideoNode alloc] init];
+  _videoNode.asset = _asset;
```

Comment:
Setting the asset here will cause the ASVideoNode to start loading the asset data asynchronously.   There are cases where the client won't perform this expensive operation when the node is initialized.  We might want to move this to when the node becomes visible.  

Or even better, we can let the client decide whether to "preload" the asset.